README.md
PyISIS

# Introduction

I have started this project around 2008 while working in collaboration of the Team from Bireme/PAHO/WHO.

This is one of the original versions.

Later, Luciano Ramalho replaced me on this project, and I bet he made lots of enhancements. But this code base has only my very humble code.

This version hosted by Gustavo Fonseca is probably more up-to-date https://bitbucket.org/gustavofonseca/pyisis/src/master/

Cheers,
Rodsenra

rsenra@acm.org

# License

This version is released under the MIT License.

# References

- http://hrueda-isis.blogspot.com/2008/06/
- https://pdfslide.net/technology/rest-gateway-e-compiladores.html



Changelog.txt
2009-11-04: João Chaves <joaochaves@gpr.com.br>
	* Restored comment block in save function;
	* Added except handling for search, search_index and listkeys functions. (tkt #115)

2009-11-03: João Chaves <joaochaves@gpr.com.br>
    * Fixed xrf offset to not extra_large header in function decode;
    * Fixed delete status in XfrRecord._write function;
    * Added routine to save "new field" to last position in save function;
    * Added auto-increase block size to mst file;
    * Added test to block save data between positions 498 and 512;
    * Fixed NXTMFP offset for incorret value;
    * Fixed BLOCK ID in xrf file;

2009-10-30: João Chaves <joaochaves@gpr.com.br>
	* Fix LEADER_SIZE in save function (records.py);
	* Removed set_hook from IPShellEmbed to engine's interactive method;
	* Removed check mask to LEADER_MASK_XL and DIR_MASK;

2009-09-17: João Chaves <joaochaves@gpr.com.br>
	* Fix list node in size function;
	* Added inverting techniques test case;
	* Added exception handler to NumFuncVal (#73);
	* Fixed NumFunc in repeatable groups handling;

2009-09-16: João Chaves <joaochaves@gpr.com.br>
	* Moved to unicodedata functions in invertdb;

2009-09-14 João Chaves <joaochaves@gpr.com.br>
	* Fix new record offset in save function;
	* Added reset flags for master file and xref file;
	* Added new test case: test_update_flags (test_files);

2009-09-11 João Chaves <joaochaves@gpr.com.br>
    * Added ref and lookup trivial syntax test case;

2009-09-03 João Chaves <joaochaves@gpr.com.br>
    * Removed hard coded ansiuctab;
    * Used unicodedata.normalize function for encode key;

2009-09-02 João Chaves <joaochaves@gpr.com.br>
	* Added ref and lookup trivial syntax;
	* Added config SEARCH_PATH option;
	* Added search_path function;

2009-09-01  João Chaves <joaochaves@gpr.com.br>
	* Added lilacs test case;

2009-08-31  João Chaves <joaochaves@gpr.com.br>
	* Fix yacc.py deprecation warnings;

2009-08-27  João Chaves <joaochaves@gpr.com.br>
    * Changed SILENT_EXCEPTIONS default value to False;
    * Fix unconditional newline (ticket #74);
    * Added fixtures folder to test cases;

2009-08-19  João Chaves <joaochaves@gpr.com.br>
    * Added checks to dependency options in pymx;
    * Fix dump_xml in module views.

2009-08-14  João Chaves <joaochaves@gpr.com.br>
	* Fix import os.path.split in mxpy.py;
	* Fix get_mfn_post function to use ansiuctab;

2009-08-13  João Chaves <joaochaves@gpr.com.br>
    * Added table ansiuctab to apply in invertdb function;
    * Added set self.pcb for empty callback option;
    * Added set fst file name if empty fst option;
	* Fix get_offset_field function;

2009-08-03 João Chaves <joaochaves@gpr.com.br>
    * Fix output encoding session (pymx);

2009-07-30 João Chaves <joaochaves@gpr.com.br>
	* Added 1000-1008 techniques;

2009-07-29 João Chaves <joaochaves@gpr.com.br>
	* Handling extra spaces in sequence class;
    * Added test case to gizmo and gsplit;
	* Fix break exception to empty workarea;
	* Fix continue exception to empty workarea;

2009-07-28 João Chaves <joaochaves@gpr.com.br>
   * Refactoring get_last_line function;

2009-07-27 João Chaves <joaochaves@gpr.com.br>
   * Fix TypeError in function sweepRepeatableLiteral;
   * Fix break_line function to single space new line;

2009-07-24 João Chaves <joaochaves@gpr.com.br>
   * Handling workarea in proc eval function;
   * Added workarea parameter in format_chain function;

2009-07-23 João Chaves <joaochaves@gpr.com.br>
   * Added cache to ref eval function;
   * Added gsplit proc eval function;
   * Added gizmo proc eval function;

2009-07-22 João Chaves <joaochaves@gpr.com.br>
   * Added cache to get_mfn_post Files Function;
   * Fixed Proc add function;
   * Refactoring mxpy handling session;
   * Revised Conditional Literal break line;
   * Revised Field format function break line;

2009-07-21 João Chaves <joaochaves@gpr.com.br>
   * Revised Field format and eval function;
   * Revised RepeatableGroup eval function;
   * Refactoring RepeatableLiteral;
   * Revised break line in Inconditional function;

2009-07-20 João Chaves <joaochaves@gpr.com.br>
   * Revised proc add regular expression;
   * Fixed break line in format function LeafNode;

2009-07-16 João Chaves <joaochaves@gpr.com.br>
   * Added suffix parameter in ConditionalLiteral parser p_suffix;

2009-07-15 João Chaves <joaochaves@gpr.com.br>
   * Fix function p_rprelit in parser.py;
   * Fix ConditionalLiteral end line;
   * Added post processing for ConditionalLiteral end line;

2009-07-14 João Chaves <joaochaves@gpr.com.br>
   * Fix filter_add proc regular expression;
   * Fix get_last_line function;
   * Fix Field eval repeatable literal;
   * Added set_InconditionalLitaralformat in BinOp class;

2009-07-13 João Chaves <joaochaves@gpr.com.br>
   * Fix handling workarea in Sequence eval function;
   * Fix LeafNode format function;

2009-07-10 João Chaves <joaochaves@gpr.com.br>
   * Fix handling RepeatableGroup workarea;
   * Fix break_line function;

2009-07-08 João Chaves <joaochaves@gpr.com.br>
   * Fix c comment regular expression in lexer.py;
   * Added support to emptycomma repeatableGroup ",(,,),";
   * Fix proc add_tag regular expression;
   * Added support to empty if statement;

2009-07-07 João Chaves <joaochaves@gpr.com.br>
   * Fixed handling lexer c style comment;
   * Fixed inconditionalLiteral new line;
   * Fixed parser p_strfunc_s funtction;
   * Fixed handling RepeatableGroup VField nodes;

2009-07-02 João Chaves <joaochaves@gpr.com.br>
   * Added repeatable literal function to set field properties;
   * Fix p_ifstatement and p_ifelsestatement in parse.py;
   * Fix handling Field in lexer:t_VFIELD function;
   * Fix handling p_rprelit and p_rposlit parser function;

2009-07-01 João Chaves <joaochaves@gpr.com.br>
   * Refactoring handling Repeatable Literal;
   * Fix VField empty production;

2009-06-30 João Chaves <joaochaves@gpr.com.br>
   * Added handling dot spacer in break_line function;


pyisis/config.py
# -*- coding: utf-8 -*-

"""
Configuration info for pyisis.
"""

__updated__ = "2008-07-22"
__created__ = "2008-04-02"
__author__  = "Rodrigo Senra <rsenra@acm.org>"

import re
import sys
import os
import logging
import tempfile
from math import log
from os.path import join, dirname, split
from struct import calcsize
from ConfigParser import ConfigParser, NoSectionError, NoOptionError
from codecs import lookup


class ConfigError(Exception):
    pass

class Config(object):
    """ This class maps the information available in a confioguration file
    (*.ini) into an instance holding configuration attributes.
    Some parameters require pre-processing to convert text from the configuration
    file into appropriate Python values.
    """
    def __init__(self):
        """Set reasonable default values. If no configuration file (.ini)
        is found later, then use these values.
        """
        self.COLLECTIONS = []

        self.INPUT_ENCODING = 'utf-8'
        self.LANGUAGE = 'en'
        # set sample collection
        if sys.platform == 'win32':
            basepath = self.basepath = join(sys.prefix, 'lib','site-packages','pyisis')
            self.OUTPUT_ENCODING = 'cp850'
        else:
            basepath = self.basepath = join(split(os.__file__)[0],'site-packages','pyisis')
            self.OUTPUT_ENCODING = 'utf-8'

        fullpath = join(basepath, 'sample')
        self.COLLECTIONS.append(('IsisCollection', 'sample', [fullpath]))

        self.FDT_LINE = r"(?P<name>\w[\w\s]*?\w)\s+(?P<subfields>\w*)\s+(?P<tag>\d*)\s+(?P<size>\d*)\s+(?P<etype>\d+)\s+(?P<repeat>\d+)"
        self.SILENT_EXCEPTIONS = False
        self.POINTER_SIZE = 4
        self.LAST_XRF_BLOCK = -1
        self.PARSER_AUXFILES_DIR = tempfile.gettempdir()
        self.YACC_DEBUG = False
        self.LOG_LEVEL = logging.DEBUG
        self.LOG_PATH = join(tempfile.gettempdir(), "pymx.log")
        self.HTTP_PORT = 8080
        self.SSH_PORT = 2222
        self.USERNAME = 'admin'
        self.PASSWORD = 'admin'
        self.BLOCK_SIZE = 512
        self.BLOCK_ALIGN = False
        self.SUBFIELD_DELIMITER = "^"
        self.CTRL_SIZE = 64
        self.BYTE_ORDER = "little" #  or "big"
        self.BYTE_ORDER_PRFIX = "<" # default standard native
        self.CTRL_MASK = self.BYTE_ORDER_PRFIX + "iiiHBBiiii"
        self.DIR_MASK = self.BYTE_ORDER_PRFIX + "HHH"
        self.LEADER_MASK = self.BYTE_ORDER_PRFIX + "iHiHHHH"
        self.LEADER_MASK_XL = self.BYTE_ORDER_PRFIX + "iHiiHHH"
        self.LEADER_XL = True
        self.XRF_BLOCK = 2048
        self.XRF_OFFSET = 0x000001FF
        self.XRF_NEW_FLAG = 0x00000400
        self.XRF_MODIFIED_FLAG = 0x00000200
        self.SEARCH_PATH = ''

        self._recalculate_dependent_params()

        # django template settings
        self.HTML_DEBUG = True
        # always use Unix style slashes
        self.HTML_TEMPLATE_DIRS = (basepath+'/web/isis/templates',)
        self.MEDIA_ROOT = basepath+'/web/isis/media'
        self.WEB_LOG = join(tempfile.gettempdir(), "isisweb.log")

        # formatting language
        self.MAX_LINE_WIDTH = 79

    def clone(self, srcobj):
        """Method to generate another Config() instance with the
        same values existing in the current self instance.
        """
        for attr in [i for i in dir(srcobj) if not i.startswith('_')]:
            value = getattr(srcobj, attr)
            if not callable(value):
                setattr(self, attr,  value)


    def _safe_set(self, attribute, cfg, section, option, validation=None):
        """Try to set attribute with cfg.get(section, option) or
        fail silently if configuration not found.
        """
        try:
            if validation is None:
                value = cfg.get(section, option)
            else:
                value = validation(cfg, section, option)
            setattr(self, attribute, value)
        except (NoSectionError, NoOptionError):
            pass

    def validate_encoding(self, cfg, section, param):
        """Validates string codecs"""
        try:
            lookup(cfg.get(section, param))
        except LookupError:
            msg = invalid_encoding%(configfile, param)
            raise ConfigError(msg)

        else:
            return value

    def validate_mask(self, cfg, section, param):
        """Validates MASK string if it is coherent with BYTE_ORDER
        and BYTE_ORDER_MASK"""
        convert = {"little":"<", "big":">"}
        byte_order_mask = cfg.get('IsisDB', 'BYTE_ORDER_MASK')
        value = cfg.get(section, param)
        if value[0]!='=' and convert[value]!=byte_order_mask:
            raise ConfigError("Inconsist byte order mask for %s and %s, expected %s"%(value, byte_order_mask, convert[value]))
        else:
            return value


    def validate_bool(self, cfg, section, param):
        """Validates boolean values converting from str to int"""
        try:
            value =  eval(cfg.get(section, param))
        except NameError:
            msg = invalid_bool%(configfile, param)
            raise ConfigError(msg)
        else:
            return value

    def validate_int(self, cfg, section, param):
        """Validates integer values converting from str to int"""
        value = cfg.get(section, param)
        try:
            value = int(value)
        except ValueError:
            msg = invalid_int%(configfile, param)
            raise ConfigError(msg)
        return value

    def _recalculate_dependent_params(self):
        """Some parameters are a function of configurable parameters.
        If the configuration changes, these values must be recalculated.
        """
        self.POINTER_PER_BLOCK = self.BLOCK_SIZE/self.POINTER_SIZE - 1
        self.BLOCK_POWER =  int(log(self.BLOCK_SIZE, 2))
        self.CTRL_MASK_SIZE = calcsize(self.CTRL_MASK)
        self.LEADER_SIZE =  calcsize(self.LEADER_MASK)
        self.LEADER_SIZE_XL =  calcsize(self.LEADER_MASK_XL)
        self.DIR_SIZE =  calcsize(self.DIR_MASK)
        if type(self.FDT_LINE)==str:
            self.FDT_LINE = re.compile(self.FDT_LINE)

    def load(self, configfile):
        """Override default values with settings from configfile"""
        cfg = ConfigParser()
        cfg.readfp(open(configfile))

        # Build collections
        try:
            for name in cfg.options("Collections"):
                value = cfg.get("Collections", name)
                # check if entry if formatted correctly
                # expects <typename>,<path>
                if "," not in value:
                    msg = invalid_collection%(configfile, value)
                    raise ConfigError(msg)

                basetype, path = [i.strip() for i in value.split(',')]

                # For the time being the config file supports only a
                # single directory per collection. However, the IsisCollection
                # constructor already accepts a list of directory paths where
                # to look for bases to include in the collection.
                # The limitation is just in the .ini and here
                self.COLLECTIONS.append((basetype, name, [path]))
        except (NoSectionError, NoOptionError):
            pass

        # Build Engine params

        # Define encodings
        self._safe_set('INPUT_ENCODING', cfg, 'Engine','INPUT_ENCODING')
        self._safe_set('OUTPUT_ENCODING', cfg, 'Engine','OUTPUT_ENCODING')
        self._safe_set('LANGUAGE', cfg, 'Engine','LANGUAGE')
        self._safe_set('FDT_LINE', cfg, 'Engine','FDT_LINE')

        # Temporary directory for parser files
        self._safe_set('PARSER_AUXFILES_DIR', cfg, 'Engine','PARSER_AUXFILES_DIR')

        # Path where to generate log files
        self._safe_set('LOG_PATH', cfg, 'Engine','LOG_PATH')

        # Search path
        self._safe_set('SEARCH_PATH', cfg, 'Engine', 'SEARCH_PATH')

        # Define log level
        try:
            self.LOG_LEVEL = logging.__dict__.get(cfg.get('Engine', 'LOG_LEVEL'),
                                              logging.DEBUG)
        except (NoOptionError,NoSectionError), ex:
            self.LOG_LEVEL = logging.DEBUG

        # Enable formatting language debugging
        self._safe_set('YACC_DEBUG', cfg, 'Engine', 'YACC_DEBUG', self.validate_bool)
        self._safe_set('SILENT_EXCEPTIONS', cfg, 'Engine', 'SILENT_EXCEPTIONS', self.validate_bool)

        # Define networking ports
        self._safe_set('HTTP_PORT', cfg, 'Gateway', 'HTTP_PORT', self.validate_int)
        self._safe_set('SSH_PORT', cfg, 'Gateway', 'SSH_PORT', self.validate_int)
        self._safe_set('USERNAME', cfg, 'Gateway', 'USERNAME')
        self._safe_set('PASSWORD', cfg, 'Gateway', 'PASSWORD')

        # Build IsisDB params
        self._safe_set('SUBFIELD_DELIMITER', cfg,'IsisDB', 'SUBFIELD_DELIMITER')
        self._safe_set('BYTE_ORDER', cfg,'IsisDB', 'BYTE_ORDER')
        self._safe_set('XRF_BLOCK', cfg, 'IsisDB', 'XRF_BLOCK', self.validate_int)
        self._safe_set('XRF_OFFSET', cfg, 'IsisDB', 'XRF_OFFSET', self.validate_int)
        self._safe_set('XRF_NEW_FLAG', cfg, 'IsisDB', 'XRF_NEW_FLAG', self.validate_int)
        self._safe_set('XRF_MODIFIED_FLAG', cfg, 'IsisDB', 'XRF_MODIFIED_FLAG', self.validate_int)
        self._safe_set('BLOCK_SIZE', cfg, 'IsisDB', 'BLOCK_SIZE', self.validate_int)
        self._safe_set('POINTER_SIZE', cfg, 'IsisDB', 'POINTER_SIZE', self.validate_int)
        self._safe_set('BLOCK_ALIGN', cfg, 'IsisDB', 'BLOCK_ALIGN', self.validate_bool)
        self._safe_set('CTRL_MASK', cfg,'IsisDB', 'CTRL_MASK', self.validate_mask)
        self._safe_set('CTRL_MASK_SIZE', cfg, 'IsisDB', 'CTRL_MASK_SIZE', self.validate_int)
        self._safe_set('CTRL_SIZE', cfg, 'IsisDB', 'CTRL_SIZE', self.validate_int)
        self._safe_set('LEADER_MASK', cfg, 'IsisDB', 'LEADER_MASK', self.validate_mask)
        self._safe_set('LEADER_SIZE', cfg, 'IsisDB', 'LEADER_SIZE', self.validate_int)
        self._safe_set('LEADER_MASK_XL', cfg, 'IsisDB', 'LEADER_MASK_XL')
        self._safe_set('LEADER_SIZE_XL', cfg, 'IsisDB', 'LEADER_SIZE_XL', self.validate_int)
        self._safe_set('LEADER_XL', cfg, 'IsisDB', 'LEADER_XL', self.validate_bool)
        self._safe_set('DIR_MASK', cfg, 'IsisDB', 'DIR_MASK')
        self._safe_set('DIR_SIZE', cfg, 'IsisDB', 'DIR_SIZE', self.validate_int)
        self._recalculate_dependent_params()

    def __str__(self):
        """Dump configuration  as string"""
        lines = []
        for name in sorted(dir(self)):
            if name.startswith("__"):
                continue # ignore hidden attributes
            value = getattr(self, name)
            if callable(value):
                continue # ignore callable attributes
            lines.append(u"%20s: %s"%(name,value))
        return "\n".join(lines)

    def __unicode__(self):
        """Dump configuration as unicode string"""
        return unicode(str(self))


# Global config instance
config = Config()

def safe_encoding(mst):
    """Given a mst reference, return its OUTPUT_ENCODING
    or the value present in the global configuration object
    """
    try:
        return mst.config.OUTPUT_ENCODING
    except AttributeError:
        return config.OUTPUT_ENCODING

